arm/Asmexpand.ml

```diff
@@ -11,8 +11,334 @@
+	 emit (Pldr (res1,addr, SOimm _4));
+	 emit (Pldr (res2,addr, SOimm _0))
+       end
+  | Mfloat32, [IR addr], [FR res] ->
+     emit (Pflds (res, addr, _0))
+  | Mfloat64, [IR addr], [FR res] ->
+     emit (Pfldd (res,addr, _0))
+  | _ ->
+     assert false
+
+let expand_builtin_vload chunk args res =
+  expand_builtin_vload_common chunk args res
+
+let expand_builtin_vload_global chunk id ofs args res =
+  emit (Ploadsymbol (IR14,id,ofs));
+  expand_builtin_vload_common chunk args res
+
+let expand_builtin_vstore_common chunk args =
+  match chunk, args with
+  | (Mint8signed | Mint8unsigned), [IR  addr; IR src] ->
+     emit (Pstrb (src,addr, SOimm _0))
+  | (Mint16signed | Mint16unsigned), [IR  addr; IR src] ->
+     emit (Pstrh (src,addr, SOimm _0))
+  | Mint32, [IR  addr; IR src] ->
+     emit (Pstr (src,addr, SOimm _0))
+  | Mint64, [IR  addr; IR src1; IR src2] ->
+     emit (Pstr (src2,addr,SOimm _0));
+     emit (Pstr (src1,addr,SOimm _0))
+  | Mfloat32, [IR  addr; FR src] ->
+     emit (Pfstd (src,addr,_0))
+  | Mfloat64, [IR  addr; FR src] ->
+     emit (Pfsts (src,addr,_0));
+  | _ ->
+     assert false
+
+let expand_builtin_vstore chunk args =
+  expand_builtin_vstore_common chunk args
+
+let expand_builtin_vstore_global chunk id ofs args =
+  emit (Ploadsymbol (IR14,id,ofs));
+  expand_builtin_vstore_common chunk args
+
+(* Handling of varargs *)
+
+let align n a = (n + a - 1) land (-a)
+
+let rec next_arg_location ir ofs = function
+  | [] ->
+     Int32.of_int (ir * 4 + ofs)
+  | (Tint | Tsingle | Tany32) :: l ->
+     if ir < 4
+     then next_arg_location (ir + 1) ofs l
+     else next_arg_location ir (ofs + 4) l
+  | (Tfloat | Tlong | Tany64) :: l ->
+     if ir < 3
+     then next_arg_location (align ir 2 + 2) ofs l
+     else next_arg_location ir (align ofs 8 + 8) l
+
+let expand_builtin_va_start r =
+  if not !current_function.fn_sig.sig_cc.cc_vararg then
+    invalid_arg "Fatal error: va_start used in non-vararg function";
+  let ofs =
+    Int32.add
+      (next_arg_location 0 0 !current_function.fn_sig.sig_args)
+      !PrintAsmaux.current_function_stacksize in
+  expand_addimm IR14 IR13 (coqint_of_camlint ofs);
+  emit (Pstr (IR14,r,SOimm _0))
+
+
+(* Handling of compiler-inlined builtins *)
+let expand_builtin_inline name args res =
+  match name, args, res with
+  (* Integer arithmetic *)
+  | ("__builtin_bswap" | "__builtin_bswap32"), [IR a1], [IR res] ->
+     emit (Prev (IR res,IR a1))
+  | "__builtin_bswap16", [IR a1], [IR res] ->
+     emit (Prev16 (IR res,IR a1))
+  | "__builtin_clz", [IR a1], [IR res] ->
+     emit (Pclz (IR res, IR a1))
+  (* Float arithmetic *)
+  | "__builtin_fabs",  [FR a1], [FR res] ->
+     emit (Pfabsd (res,a1))
+  | "__builtin_fsqrt", [FR a1], [FR res] ->
+     emit (Pfsqrt (res,a1))
+  (* 64-bit integer arithmetic *)
+  | "__builtin_negl", [IR ah;IR al], [IR rh; IR rl] ->
+     emit (Prsbs (rl,al,SOimm _0));
+     if !Clflags.option_mthumb then begin
+	 emit (Pmvn (rh,SOreg ah));
```

Comment:
There was a comment here in the original code that explains the magic and should be left.